vm/ubpf_instruction_valid.c

```diff
@@ -0,0 +1,1010 @@
+    },
+    {
+        .opcode = EBPF_OP_ATOMIC32_STORE,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_10,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+        .immediate_lower_bound = 0x0,
+        .immediate_upper_bound = 0xff,
+        .offset_lower_bound = INT16_MIN,
+        .offset_upper_bound = INT16_MAX,
+    },
+    {
+        .opcode = EBPF_OP_ATOMIC_STORE,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_10,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+        .immediate_lower_bound = 0x0,
+        .immediate_upper_bound = 0xff,
+        .offset_lower_bound = INT16_MIN,
+        .offset_upper_bound = INT16_MAX,
+    },
+};
+
+static ubpf_inst_filter_t* _ubpf_filter_instruction_lookup_table[256];
+
+/**
+ * @brief Initialize the lookup table for the instruction filter.
+ */
+static void _initialize_lookup_table()
+{
+    static bool _initialized = false;
+
+    if (_initialized) {
+        return;
+    }
+
+    for (int i = 0; i < sizeof(_ubpf_instruction_filter) / sizeof(_ubpf_instruction_filter[0]); i++) {
+        _ubpf_filter_instruction_lookup_table[_ubpf_instruction_filter[i].opcode] = &_ubpf_instruction_filter[i];
+    }
+
+    _initialized = true;
+}
+
+
+static bool _in_range(int value, int lower_bound, int upper_bound)
+{
+    return value >= lower_bound && value <= upper_bound;
+}
+
+bool
+ubpf_is_valid_instruction(const struct ebpf_inst insts, char ** errmsg)
+{
+    _initialize_lookup_table();
+
+    // Lookup the instruction.
+    ubpf_inst_filter_t* filter = _ubpf_filter_instruction_lookup_table[insts.opcode];
+
+    if (filter == NULL) {
+        *errmsg = ubpf_error("Invalid instruction opcode %2X.", insts.opcode);
+        return false;
+    }
+
+    // Validate the instruction.
+
+    // Validate destination register.
+    if (!_in_range(insts.dst, filter->destination_lower_bound, filter->destination_upper_bound)) {
+        *errmsg = ubpf_error("Invalid destination register %d for opcode %2X.", insts.dst, insts.opcode);
+        return false;
+    }
+
+    // Validate source register.
+    if (!_in_range(insts.src, filter->source_lower_bound, filter->source_upper_bound)) {
+        *errmsg = ubpf_error("Invalid source register %d for opcode %2X.", insts.src, insts.opcode);
+        return false;
+    }
+
+    // Validate immediate value.
+    if (!_in_range(insts.imm, filter->immediate_lower_bound, filter->immediate_upper_bound)) {
+        *errmsg = ubpf_error("Invalid immediate value %d for opcode %2X.", insts.imm, insts.opcode);
+        return false;
+    }
+
+    // Validate offset value.
+    if (!_in_range(insts.offset, filter->offset_lower_bound, filter->offset_upper_bound)) {
+        *errmsg = ubpf_error("Invalid offset value %d for opcode %2X.", insts.offset, insts.opcode);
+        return false;
+    }
+
```

Comment:
Are we able to validate things that are within the opcode and need to have specific values? In particular, the ALU64 endianness instruction is defined such that source operand field is 0: https://www.ietf.org/archive/id/draft-ietf-bpf-isa-04.html#section-4.2-3